flare/kernels/three_body_mc_simple.py

```diff
@@ -0,0 +1,1263 @@
+        cross_bond_inds_2 (np.ndarray): Two dimensional array whose row m
+            contains the indices of atoms n > m in the second local
+            environment that are within a distance r_cut of both atom n and
+            the central atom.
+        cross_bond_dists_1 (np.ndarray): Two dimensional array whose row m
+            contains the distances from atom m of atoms n > m in the first
+            local environment that are within a distance r_cut of both atom
+            n and the central atom.
+        cross_bond_dists_2 (np.ndarray): Two dimensional array whose row m
+            contains the distances from atom m of atoms n > m in the second
+            local environment that are within a distance r_cut of both atom
+            n and the central atom.
+        triplets_1 (np.ndarray): One dimensional array of integers whose entry
+            m is the number of atoms in the first local environment that are
+            within a distance r_cut of atom m.
+        triplets_2 (np.ndarray): One dimensional array of integers whose entry
+            m is the number of atoms in the second local environment that are
+            within a distance r_cut of atom m.
+        sig (float): 3-body signal variance hyperparameter.
+        ls (float): 3-body length scale hyperparameter.
+        r_cut (float): 3-body cutoff radius.
+        cutoff_func (Callable): Cutoff function.
+
+    Returns:
+        float:
+            Value of the 3-body force/energy kernel.
+    """
+    kern = np.zeros(6)
+
+    sig2 = sig * sig
+    ls1 = 1 / (2 * ls * ls)
+    ls2 = 1 / (ls * ls)
+
+    for m in range(bond_array_1.shape[0]):
+        ri1 = bond_array_1[m, 0]
+        ei1 = etypes1[m]
+
+        for n in range(triplets_1[m]):
+            ind1 = cross_bond_inds_1[m, m + n + 1]
+            ri2 = bond_array_1[ind1, 0]
+            ei2 = etypes1[ind1]
+
+            ri3 = cross_bond_dists_1[m, m + n + 1]
+            fi3, _ = cutoff_func(r_cut, ri3, 0)
+
+            for p in range(bond_array_2.shape[0]):
+                rj1 = bond_array_2[p, 0]
+                fj1, _ = cutoff_func(r_cut, rj1, 0)
+                ej1 = etypes2[p]
+
+                for q in range(triplets_2[p]):
+                    ind2 = cross_bond_inds_2[p, p + q + 1]
+                    rj2 = bond_array_2[ind2, 0]
+                    fj2, _ = cutoff_func(r_cut, rj2, 0)
+                    ej2 = etypes2[ind2]
+                    rj3 = cross_bond_dists_2[p, p + q + 1]
+                    fj3, _ = cutoff_func(r_cut, rj3, 0)
+                    fj = fj1 * fj2 * fj3
+
+                    r11 = ri1 - rj1
+                    r12 = ri1 - rj2
+                    r13 = ri1 - rj3
+                    r21 = ri2 - rj1
+                    r22 = ri2 - rj2
+                    r23 = ri2 - rj3
+                    r31 = ri3 - rj1
+                    r32 = ri3 - rj2
+                    r33 = ri3 - rj3
+
+                    stress_count = 0
+                    for d1 in range(3):
```

Comment:
data locality would be better if the d1 and d2 loop is moved to the right before the helper function and let ci1, ci2 to be 1d vector. so fi and fj won't be computed for 9 times.

Reply:
Addressed in 6952612